src/libcore/iter/adapters/mod.rs

```diff
@@ -1433,6 +1497,24 @@ impl<I: Iterator> Peekable<I> {
     }
 }
 
+#[unstable(feature = "trusted_len", issue = "37572")]
+unsafe impl<I> TrustedLen for Peekable<I>
+    where I: TrustedLen,
+{}
+
+#[unstable(issue = "0", feature = "inplace_iteration")]
+impl<S: Iterator, I: Iterator> SourceIter for Peekable<I> where I: SourceIter<Source = S> {
+    type Source = S;
+
+    #[inline]
+    fn as_inner(&mut self) -> &mut S {
+        SourceIter::as_inner(&mut self.iter)
+    }
+}
+
+#[unstable(issue = "0", feature = "inplace_iteration")]
+unsafe impl<I: InPlaceIterable> InPlaceIterable for Peekable<I> {}
```

Comment:
Peekable does not appear to uphold the required contract of InPlaceIterable. In particular when there is a buffered "peeked" item, calling next will not advance the underlying iterator.

Reply:
That only means the underlying iterator will have been advanced... in advance.